### PR TITLE
Mirroring script: Combine statements in array

### PR DESCRIPTION
**Pull request by queengooborg**
This PR is a subset from #6582 that introduces a function to combine statements in an array as needed when mirroring.

### CODE REVIEW

File: scripts/mirror.js [hunk 1]
```diff
@@ -144,9 +144,7 @@ const combineNotes = (notes1, notes2) => {
     }
   }
 
-  newNotes = newNotes.filter((item, pos) => {
-    newNotes.indexOf(item) == pos;
-  });
+  newNotes = newNotes.filter((item, pos) => newNotes.indexOf(item) == pos);
```
Comment (caugner) [inline scripts/mirror.js:147]: Does this deduplicate the notes? And wasn't the old implementation broken, i.e. the block returned `undefined` for every note?

File: scripts/mirror.js [hunk 2]
```diff
@@ -193,6 +191,79 @@ const copyStatement = (data) => {
   return newData;
 };
 
+/**
+ * @param {...SupportStatement} data
+ * @returns {SupportStatement}
+ */
+const combineStatements = (...data) => {
+  const ignored_keys = ['version_added', 'notes'];
```
Comment (caugner) [inline scripts/mirror.js:199]: CamelCase for consistency:

```suggestion
  const ignoredKeys = ['version_added', 'notes'];
```

File: scripts/mirror.js [hunk 3]
```diff
@@ -193,6 +191,79 @@ const copyStatement = (data) => {
...
+
+  let flattenedData = data.flat(2);
+  let sections = {};
+  let newData = [];
+
+  for (const d of flattenedData) {
+    let key = Object.keys(d)
```
Comment (caugner) [inline scripts/mirror.js:206]:
```suggestion
    const key = Object.keys(d)
```

File: scripts/mirror.js [hunk 4]
```diff
@@ -193,6 +191,79 @@ const copyStatement = (data) => {
...
+            currentVA.replace('≤', ''),
+            newVA.replace('≤', ''),
+            '>',
+          )
+        ) {
+          currentStatement.version_added = newVA;
+        }
+      }
+
+      let newNotes = combineNotes(currentStatement.notes, newStatement.notes);
```
Comment (caugner) [inline scripts/mirror.js:244]:
```suggestion
      const newNotes = combineNotes(currentStatement.notes, newStatement.notes);
```

File: scripts/mirror.js [hunk 5]
```diff
@@ -193,6 +191,79 @@ const copyStatement = (data) => {
...
+      newData.push(currentStatement);
+      continue;
+    }
+
+    for (const i in sections[k]) {
+      if (i === 0) continue;
+      let newStatement = sections[k][i];
+
+      let currentVA = currentStatement.version_added;
+      let newVA = newStatement.version_added;
```
Comment (caugner) [inline scripts/mirror.js:226]:
```suggestion
      const currentVA = currentStatement.version_added;
      const newVA = newStatement.version_added;
```

File: scripts/mirror.js [hunk 6]
```diff
@@ -193,6 +191,79 @@ const copyStatement = (data) => {
...
+    let currentStatement = sections[k][0];
+
+    if (sections[k].length == 1) {
+      newData.push(currentStatement);
+      continue;
+    }
+
+    for (const i in sections[k]) {
+      if (i === 0) continue;
+      let newStatement = sections[k][i];
```
Comment (caugner) [inline scripts/mirror.js:223]:
```suggestion
      const newStatement = sections[k][i];
```

File: scripts/mirror.js [hunk 7]
```diff
@@ -193,6 +191,79 @@ const copyStatement = (data) => {
...
+  for (const d of flattenedData) {
+    let key = Object.keys(d)
+      .filter((k) => !ignored_keys.includes(k))
+      .join('');
+    if (!(key in sections)) sections[key] = [];
+    sections[key].push(d);
+  }
+
+  for (const k of Object.keys(sections)) {
+    let currentStatement = sections[k][0];
```
Comment (caugner) [inline scripts/mirror.js:214]:
```suggestion
    const currentStatement = sections[k][0];
```

File: scripts/mirror.js [hunk 8]
```diff
@@ -193,6 +191,79 @@ const copyStatement = (data) => {
...
+/**
+ * @param {...SupportStatement} data
+ * @returns {SupportStatement}
+ */
+const combineStatements = (...data) => {
+  const ignored_keys = ['version_added', 'notes'];
+
+  let flattenedData = data.flat(2);
+  let sections = {};
+  let newData = [];
```
Comment (caugner) [inline scripts/mirror.js:203]:
```suggestion
  const flattenedData = data.flat(2);
  const sections = {};
  const newData = [];
```

File: scripts/mirror.js [hunk 9]
```diff
@@ -193,6 +191,79 @@ const copyStatement = (data) => {
...
+      let newNotes = combineNotes(currentStatement.notes, newStatement.notes);
+      if (newNotes) currentStatement.notes = newNotes;
+    }
+
+    if ('notes' in currentStatement && !currentStatement.notes)
+      delete currentStatement.notes;
+    newData.push(currentStatement);
+  }
+
+  if (newData.length === 1) return newData[0];
```
Comment (caugner) [inline scripts/mirror.js:253]: Unconfuse:

```suggestion
    if ('notes' in currentStatement && !currentStatement.notes) {
      delete currentStatement.notes;
    }
    newData.push(currentStatement);
  }

  if (newData.length === 1) {
    return newData[0];
  }
```

File: scripts/mirror.js [hunk 10]
```diff
@@ -428,15 +499,11 @@ const bumpVersion = (sourceData, originalData, destination, targetVersion) => {
...
-      newData[i] = bumpVersion(
-        sourceData[i],
-        originalData,
-        destination,
-        targetVersion,
-      );
-    }
+    newData = combineStatements(
+      ...sourceData.map((d) =>
+        bumpVersion(d, originalData, destination, targetVersion),
```
Comment (caugner) [inline scripts/mirror.js:504]: I'd avoid single-char variables:

```suggestion
      ...sourceData.map((data) =>
        bumpVersion(data, originalData, destination, targetVersion),
```

**Review by caugner — CHANGES_REQUESTED** [review]

File: scripts/mirror.js (hunk 1 above)
Comment (queengooborg) [reply scripts/mirror.js:147]: That's correct, for both!

**Review by caugner — APPROVED** [review]
LGTM

**Pull request merged**